### PR TITLE
Refactor base transform

### PR DESCRIPTION
**Pull request by ly015**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Refactor base transform and transform wrappers.

### `cache_randomness` 
1. Rename decorator `cacheable_method` to `cache_randomness`
1. Allow decorated method not to be invoked in transform (but invoking multiple times is not allowed)

### `Remap` -> `KeyMapper`
Modify parameters:
1.  `transforms`: Allow single transform, transform list, and None (no transform)
1. `input_mapping` -> `mapping`
1. `output_mapping` -> `remapping`
1. `inplace` -> `auto_remap`: Default value is None, which means determine the bool value from `remapping`. If `remapping` is None, set `auto_remap=True`, vice versa.
1. `strict` -> `allow_nonexist_keys`: Default: False
1. remove the unnecessary requirement that `all_nonexist_keys` must be False when `auto_remap` is True. 

### `ApplyToMultiple` -> `TransformBroadcaster`
Parameters modified according to `KeyMapper`

### `RandomChoice`
Modify parameters:
1. `pipelines` -> `transforms`
1. `pipeline_probs` -> `prob`

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: docs/zh_cn/understand_mmcv/data_transform.md [hunk 1]
```diff
@@ -192,15 +192,15 @@ pipeline = [
...
+        prob=[0.4, 0.6]  # 两种随机变换组合各自的选用概率
         )
     ...
 ]
 ```
 
 ### 多目标扩展（ApplyToMultiple）
 
-通常，一个数据变换类只会从一个固定的字段读取操作目标。虽然我们也可以使用 `Remap` 来改变读取的字段，但无法将变换一次性应用于多个字段的数据。为了实现这一功能，我们需要借助多目标扩展包装（`ApplyToMultiple`）。
+通常，一个数据变换类只会从一个固定的字段读取操作目标。虽然我们也可以使用 `ApplyToMapped` 来改变读取的字段，但无法将变换一次性应用于多个字段的数据。为了实现这一功能，我们需要借助多目标扩展包装（`ApplyToMultiple`）。
```
Comment (gaotongxiao) [inline docs/zh_cn/understand_mmcv/data_transform.md:203]: It's really hard for me to grasp the meaning of `ApplyToMapped` without reading the documentation. I'd suggest `KeyMapper`, which is more clear and emphasizes its role as a helper/wrapper by the name itself.

File: docs/zh_cn/understand_mmcv/data_transform.md [hunk 2]
```diff
@@ -192,15 +192,15 @@ pipeline = [
...
 
 多目标扩展包装（`ApplyToMultiple`）有两个用法，一是将数据变换作用于指定的多个字段，二是将数据变换作用于某个字段下的一组目标中。
```
Comment (gaotongxiao) [inline docs/zh_cn/understand_mmcv/data_transform.md:205]: Proposing `TransformBrocaster` here. The role of this transform also seems clearer though the name is a bit hard to pronounce. (I find that ROS has approved this name in their library http://docs.ros.org/en/hydro/api/tf/html/c++/classtf_1_1TransformBroadcaster.html)

File: mmcv/transforms/utils.py [hunk 3]
```diff
@@ -11,29 +11,29 @@
...
 
     This decorator is usually used together with the context-manager
-    :func`:cache_random_params`. In this context, a cacheable method will
+    :func`:cache_random_params`. In this context, a decorated method will
     cache its return value(s) at the first time of being invoked, and always
     return the cached values when being invoked again.
 
     .. note::
-        Only a instance method can be decorated as a cacheable_method.
+        Only a instance method can be decorated as a cache_randomness.
```
Comment (gaotongxiao) [inline mmcv/transforms/utils.py:24]:
```suggestion
        Only an instance method can be decorated by ``cache_randomness``.
```

File: mmcv/transforms/utils.py [hunk 4]
```diff
@@ -81,13 +81,13 @@ def __get__(self, obj, cls):
 
 @contextmanager
 def cache_random_params(transforms: Union[BaseTransform, Iterable]):
-    """Context-manager that enables the cache of cacheable methods in
-    transforms.
+    """Context-manager that enables the cache of return values of methods
+    decorated by `cache_randomness` in transforms.
```
Comment (gaotongxiao) [inline mmcv/transforms/utils.py:85]:
```suggestion
    decorated by ``cache_randomness`` in transforms.
```
Double backquotes for code samples

File: mmcv/transforms/utils.py [hunk 5]
```diff
@@ -81,13 +81,13 @@ def __get__(self, obj, cls):
...
 
-    In this mode, cacheable methods will cache their return values on the
+    In this mode, decorated methods will cache their return values on the
     first invoking, and always return the cached value afterward. This allow
     to apply random transforms in a deterministic way. For example, apply same
-    transforms on multiple examples. See `cacheable_method` for more
+    transforms on multiple examples. See `cache_randomness` for more
```
Comment (gaotongxiao) [inline mmcv/transforms/utils.py:90]:
```suggestion
    transforms on multiple examples. See ``cache_randomness`` for more
```

File: mmcv/transforms/wrappers.py [hunk 6]
```diff
@@ -148,66 +155,72 @@ class Remap(BaseTransform):
...
+
+        if auto_remap is None:
+            auto_remap = remapping is None
+        self.auto_remap = auto_remap
+
+        if self.auto_remap:
+            if self.allow_nonexist_keys:
+                raise ValueError(
+                    'ApplyToMapped: `allow_nonexist_keys` must be set False if'
+                    '`auto_remap` is set True.')
```
Comment (mzr1996) [inline mmcv/transforms/wrappers.py:175]: Unncessary requirement

File: docs/zh_cn/understand_mmcv/data_transform.md (hunk 1 above)
Comment (ZwwWayne) [reply docs/zh_cn/understand_mmcv/data_transform.md:203]: I think its simpler, let's try `KeyMapper`.

File: docs/zh_cn/understand_mmcv/data_transform.md (hunk 2 above)
Comment (ly015) [reply docs/zh_cn/understand_mmcv/data_transform.md:205]: @gaotongxiao @ZwwWayne 
Which one would be better, `KeyMapper` and `TransformBroadcaster` or `KeyMap` and `TransformBroadcast`？ Do we have a convention about transform names being verbal or nominal?

Comment (gaotongxiao) [reply docs/zh_cn/understand_mmcv/data_transform.md:205]: Adapting a different name convention for transform wrappers may help users tell them apart from other transforms, so I'd vote for the names ending with "-er". Also, `RandomChoice` -> `RandomChooser`

Comment (mzr1996) [reply docs/zh_cn/understand_mmcv/data_transform.md:205]: > Using a different name convention for transform wrappers may help users tell them apart from other transforms, so I'd vote for the names ending with "-er". Also, `RandomChoice` -> `RandomChooser`

But `RandomChoice` is align with [`RandomChoice` in torchvision](https://pytorch.org/vision/stable/generated/torchvision.transforms.RandomChoice.html#torchvision.transforms.RandomChoice)

Comment (gaotongxiao) [reply docs/zh_cn/understand_mmcv/data_transform.md:205]: I see, we can stick to `RandomChoice`

**Review by gaotongxiao — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**